File: adgeistkit/src/main/java/com/adgeistkit/ads/BaseAdView.java
package com.adgeistkit.ads;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresPermission;

import com.adgeistkit.AdgeistCore;
import com.adgeistkit.R;
import com.adgeistkit.data.models.Creative;
import com.adgeistkit.data.models.Dimensions;
import com.adgeistkit.data.models.DisplayOptions;
import com.adgeistkit.data.models.FixedAdResponse;
import com.adgeistkit.data.network.FetchCreative;
import com.google.gson.Gson;

import java.util.*;

public abstract class BaseAdView extends ViewGroup {
    private static final String TAG = "BaseAdView";
    
    @Nullable
    protected AdSize adSize;
    
    @NonNull
    protected String adUnitId = "";

    @Nullable
    private WebView webView;
    
    private boolean isLoading = false;
    private boolean isDestroyed = false;
    private Handler mainHandler;

    @Nullable
    public AdSize getAdSize() {
        return adSize;
    }

    @NonNull
    public String getAdUnitId() {
        return adUnitId;
    }

    protected BaseAdView(@NonNull Context context, int adViewType) {
        super(context);
        initialize(context, null);
    }

    protected BaseAdView(@NonNull Context context, @NonNull AttributeSet attrs, int adViewType) {
        super(context, attrs);
        initialize(context, attrs);
    }

    protected BaseAdView(@NonNull Context context, @NonNull AttributeSet attrs, int defStyle, int adViewType) {
        super(context, attrs, defStyle);
        initialize(context, attrs);
    }

    private void initialize(@NonNull Context context, @Nullable AttributeSet attrs) {
        mainHandler = new Handler(Looper.getMainLooper());
        
        if (attrs != null) {
            TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.AdView);
            try {
                // Parse adUnitId from XML
                String xmlAdUnitId = typedArray.getString(R.styleable.AdView_adUnitId);
                if (xmlAdUnitId != null && !xmlAdUnitId.isEmpty()) {
                    adUnitId = xmlAdUnitId;
                }
                
                // Parse adSize from XML
                int sizeIndex = typedArray.getInt(R.styleable.AdView_adSize, -1);
                adSize = getAdSizeFromIndex(sizeIndex);
            } finally {
                typedArray.recycle();
            }
        }
        
        // Default ad size if not set
        if (adSize == null) {
            adSize = AdSize.BANNER;
        }
    }

    private AdSize getAdSizeFromIndex(int index) {
        switch (index) {
            case 0: return AdSize.BANNER;
            case 1: return AdSize.LARGE_BANNER;
            case 2: return AdSize.MEDIUM_RECTANGLE;
            case 3: return AdSize.FULL_BANNER;
            case 4: return AdSize.LEADERBOARD;
            case 5: return AdSize.WIDE_SKYSCRAPER;
            default: return AdSize.BANNER;
        }
    }

    private String getMetaValue(Context context, String key) {
        try {
            ApplicationInfo ai = context.getPackageManager()
                    .getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);

            Bundle bundle = ai.metaData;
            return bundle != null ? bundle.getString(key) : null;

        } catch (Exception e) {
            Log.e("BaseAdView", "Meta-data read failed for: " + key, e);
            return null;
        }
    }

    @RequiresPermission("android.permission.INTERNET")
    public void loadAd(@NonNull AdRequest adRequest) {
        if (isDestroyed) {
            Log.e(TAG, "Cannot load ad - AdView has been destroyed");
            return;
        }
        
        if (isLoading) {
            Log.w(TAG, "Ad is already loading");
            return;
        }
        
        if (adUnitId == null || adUnitId.isEmpty()) {
            Log.e(TAG, "Ad unit ID is null or empty");
            return;
        }
        
        isLoading = true;

        new Thread(() -> {
            try {
                AdgeistCore adgeist = AdgeistCore.getInstance();
                FetchCreative fetchCreative = adgeist.getCreative();

                String publisherId = getMetaValue(getContext(), "com.adgeistkit.ads.APP_ID");
                String apiKey = getMetaValue(getContext(), "com.adgeistkit.ads.API_KEY");

                fetchCreative.fetchCreative(
                    apiKey,"https://adgeist-ad-integration.d49kd6luw1c4m.amplifyapp.com/", adUnitId, publisherId, "FIXED", true,
                    creativeData -> {
                        mainHandler.post(() -> {
                            isLoading = false;
                            if (creativeData == null) {
                                return;
                            }
                            try {

                                FixedAdResponse fixed = (FixedAdResponse) creativeData;
                                if (fixed == null) {
                                    return;
                                }
                                Creative c = fixed.getCreatives().get(0);

                                Map<String, Object> simpleCreative = new HashMap<>();

                                simpleCreative.put("adElementId", "adgeist_ads_iframe_" + adUnitId);
                                simpleCreative.put("title", c.getTitle());
                                simpleCreative.put("description", c.getDescription());
                                simpleCreative.put("name", fixed.getAdvertiser() != null ? fixed.getAdvertiser().getName() : "-");
                                simpleCreative.put("ctaUrl", c.getCtaUrl());
                                simpleCreative.put("fileUrl", c.getFileUrl());
                                simpleCreative.put("type", c.getType());

                                DisplayOptions options = fixed.getDisplayOptions();
                                boolean isResponsive = options != null && options.isResponsive() != null ? options.isResponsive() : false;
                                String responsiveType = options != null && options.getResponsiveType() != null ? options.getResponsiveType() : "Square";

                                simpleCreative.put("isResponsive", isResponsive);
                                simpleCreative.put("responsiveType", responsiveType);

                                Dimensions dim = options != null ? options.getDimensions() : null;
                                int width = dim != null && dim.getWidth() != null ? dim.getWidth() : 300;
                                int height = dim != null && dim.getHeight() != null ? dim.getHeight() : 300;

                                simpleCreative.put("width", width);
                                simpleCreative.put("height", height);

                                simpleCreative.put("adspaceType", "banner");

                                List<Map<String, String>> mediaList = new ArrayList<>();

                                if (c.getFileUrl() != null) {
                                    Map<String, String> mediaObj = new HashMap<>();
                                    mediaObj.put("src", c.getFileUrl());
                                    mediaList.add(mediaObj);
                                }

                                simpleCreative.put("media", mediaList);

                                String mediaType = (c.getType() != null) ? c.getType() : "image";
                                simpleCreative.put("mediaType", mediaType);

                                String creativeJson = new Gson().toJson(simpleCreative);

                                if (creativeJson != null) {
                                    renderAdWithAdCard(creativeJson);
                                    notifyAdLoaded();
                                }
                            } catch (Exception err) {
                                Log.e(TAG, "Error parsing creativeData", err);
                            }

                        });

                        return null;
                    }
                );
            } catch (Exception e) {
                mainHandler.post(() -> {
                    isLoading = false;
                    Log.e(TAG, "Error loading ad", e);
                });
            }
        }).start();
    }

    private void renderAdWithAdCard(@NonNull String creativeJsonData) {
        removeAllViews();
        
        webView = new WebView(getContext());
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setDomStorageEnabled(true);
        webView.getSettings().setLoadWithOverviewMode(true);
        webView.getSettings().setUseWideViewPort(true);

        // Enable WebView debugging (you can inspect in Chrome DevTools)
        // chrome://inspect/#devices
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(true);
        }
        
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                Log.i(TAG, "✅ WebView page finished loading: " + url);
            }

            @Override
            public void onLoadResource(WebView view, String url) {
                super.onLoadResource(view, url);
                Log.d(TAG, "📦 Loading resource: " + url);
            }
        });
        
        // Set WebChromeClient to capture console logs
        webView.setWebChromeClient(new android.webkit.WebChromeClient() {
            @Override
            public boolean onConsoleMessage(android.webkit.ConsoleMessage consoleMessage) {
                String logLevel = consoleMessage.messageLevel().name();
                String message = consoleMessage.message();
                String source = consoleMessage.sourceId();
                int line = consoleMessage.lineNumber();
                
                // Format: [LEVEL] message (source:line)
                String fullLog = String.format("[%s] %s (%s:%d)", 
                    logLevel, message, source, line);
                Log.e(TAG, fullLog);
                // Log based on level
                switch (consoleMessage.messageLevel()) {
                    case ERROR:
                        Log.e(TAG, "🔴 JS Error: " + fullLog);
                        break;
                    case WARNING:
                        Log.w(TAG, "🟡 JS Warning: " + fullLog);
                        break;
                    case DEBUG:
                    case LOG:
                    case TIP:
                    default:
                        Log.d(TAG, "🔵 JS Log: " + fullLog);
                        break;
                }
                
                return true;
            }
        });
        
        // Set click listener
        webView.setOnClickListener(v -> {
        });
        
        // Build HTML that loads AdCard.js and renders the creative
        String htmlContent = buildAdCardHtml(creativeJsonData);

        webView.loadDataWithBaseURL(
            "https://adgeist.ai",  
            htmlContent,        
            "text/html",         
            "UTF-8",           
            null           
        );
        
        // Add to view hierarchy
        addView(webView, new LayoutParams(
            LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT
        ));
    }

    private String buildAdCardHtml(@NonNull String creativeJsonData) {
        String escapedJson = creativeJsonData
                .replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r");

        Log.i(TAG, creativeJsonData);

        return "<!DOCTYPE html>" +
                "<html>" +
                "<head>" +
                "    <meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no'>" +
                "    <style>" +
                "        * {" +
                "            margin: 0;" +
                "            padding: 0;" +
                "            box-sizing: border-box;" +
                "        }" +
                "        html, body {" +
                "            width: 100%;" +
                "            height: 100%;" +
                "            margin: 0;" +
                "            padding: 0;" +
                "            overflow: hidden;" +
                "        }" +
                "        body {" +
                "            display: flex;" +
                "            justify-content: center;" +
                "            align-items: center;" +
                "        }" +
                "        #ad-content {" +
                "            width: 100%;" +
                "            height: 100%;" +
                "        }" +
                "        #ad-content > * {" +
                "            width: 100%;" +
                "            height: 100%;" +
                "            object-fit: inherit;" +
                "        }" +
                "    </style>" +
                "</head>" +
                "<body>" +
                "    <div id='ad-content'></div>" +
                "    <!-- Load AdCard.js library from S3 -->" +
                "    <script src='https://adserv-scripts.s3.ap-south-1.amazonaws.com/adcard.js'></script>" +
                "    <script>" +
                "        function initAd() {" +
                "            if (typeof AdCard === 'undefined') {" +
                "                console.error('AdCard library not loaded');" +
                "                return;" +
                "            }" +
                "            try {" +
                "                const creativeData = JSON.parse(\"" + escapedJson + "\");" +
                "                const adCard = new AdCard(creativeData);" +
                "                const html = adCard.renderHtml();" +
                "                document.getElementById('ad-content').innerHTML = html;" +
                "                document.getElementById('ad-content').addEventListener('click', function(e) {" +
                "                    console.log('Ad clicked');" +
                "                });" +
                "            } catch (error) {" +
                "                console.error('Error rendering ad:', error);" +
                "            }" +
                "        }" +
                "        if (document.readyState === 'loading') {" +
                "            document.addEventListener('DOMContentLoaded', function() {" +
                "                setTimeout(initAd, 100);" +
                "            });" +
                "        } else {" +
                "            setTimeout(initAd, 100);" +
                "        }" +
                "    </script>" +
                "</body>" +
                "</html>";
    }

    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        View child = getChildAt(0);
        if (child != null && child.getVisibility() != View.GONE) {
            int width = child.getMeasuredWidth();
            int height = child.getMeasuredHeight();
            
            int horizontalSpacing = (right - left - width) / 2;
            int verticalSpacing = (bottom - top - height) / 2;
            
            child.layout(
                horizontalSpacing,
                verticalSpacing,
                horizontalSpacing + width,
                verticalSpacing + height
            );
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        View child = getChildAt(0);
        int width;
        int height;
        
        if (child != null && child.getVisibility() != View.GONE) {
            measureChild(child, widthMeasureSpec, heightMeasureSpec);
            width = child.getMeasuredWidth();
            height = child.getMeasuredHeight();
        } else {
            if (adSize != null) {
                width = adSize.getWidthInPixels(getContext());
                height = adSize.getHeightInPixels(getContext());
            } else {
                width = 0;
                height = 0;
            }
        }
        
        width = Math.max(width, getSuggestedMinimumWidth());
        height = Math.max(height, getSuggestedMinimumHeight());
        
        setMeasuredDimension(
            resolveSize(width, widthMeasureSpec),
            resolveSize(height, heightMeasureSpec)
        );
    }

    public void setAdSize(@NonNull AdSize adSize) {
        if (adSize == null) {
            throw new IllegalArgumentException("AdSize cannot be null");
        }
        this.adSize = adSize;
        requestLayout();
    }

    public void setAdUnitId(@NonNull String adUnitId) {
        if (adUnitId == null) {
            throw new IllegalArgumentException("Ad unit ID cannot be null");
        }
        this.adUnitId = adUnitId;
    }


    public boolean isCollapsible() {
        return false;
    }

    public boolean isLoading() {
        return isLoading;
    }

    private void notifyAdLoaded() {

    }
}

File: adgeistkit/src/main/java/com/adgeistkit/ads/BaseAdView.kt
package com.adgeistkit.ads

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.webkit.ConsoleMessage
import android.webkit.ConsoleMessage.MessageLevel
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.annotation.RequiresPermission
import com.adgeistkit.AdgeistCore.Companion.getInstance
import com.adgeistkit.R
import com.adgeistkit.request.AdRequest
import com.adgeistkit.data.models.FixedAdResponse
import com.adgeistkit.data.network.FetchCreative
import com.adgeistkit.logging.SdkShield
import com.google.gson.Gson
import kotlin.math.max

open class BaseAdView : ViewGroup {
    companion object {
        private const val TAG = "BaseAdView"
    }

    /**
     * Required parameters for ad rendering configuration
     */
    var adSize: AdSize? = null
    var adUnitId: String = ""
    var adType: AdType = AdType.BANNER
    var adIsResponsive: Boolean = false
    var isTestMode: Boolean = false

    /**
     * Metadata and media type for ad tracking
     */
    var metaData: String = ""
    var mediaType: String? = null

    /**
     * WebView and JavaScript bridge instances
     */
    internal var webView: WebView? = null
    private var jsInterface: JsBridge? = null

    /**
     * Listener for ad lifecycle events
     */
    var listener: AdListener? = null

    private var isLoading: Boolean = false
    private var isDestroyed = false
    private var mainHandler: Handler? = null

    protected constructor(context: Context, adViewType: Int) : super(context) {
        initialize(context, null)
    }

    protected constructor(context: Context, attrs: AttributeSet, adViewType: Int) : super(
        context,
        attrs
    ) {
        initialize(context, attrs)
    }

    protected constructor(
        context: Context,
        attrs: AttributeSet,
        defStyle: Int,
        adViewType: Int
    ) : super(context, attrs, defStyle) {
        initialize(context, attrs)
    }

    /**
     * Initializes the BaseAdView with context and attributes.
     * Sets up the main thread handler and parses XML attributes if provided.
     *
     * @param context The Android context
     * @param attrs AttributeSet from XML layout (nullable)
     */
    private fun initialize(context: Context, attrs: AttributeSet?) {
        mainHandler = Handler(Looper.getMainLooper())

        if (attrs != null) {
            val typedArray = context.obtainStyledAttributes(attrs, R.styleable.AdView)
            try {
                // Parse adUnitId from XML
                val xmlAdUnitId = typedArray.getString(R.styleable.AdView_adUnitId)
                if (xmlAdUnitId != null && !xmlAdUnitId.isEmpty()) {
                    adUnitId = xmlAdUnitId
                }
            } finally {
                typedArray.recycle()
            }
        }
    }

    /**
     * Sets the ad listener to receive ad lifecycle events.
     *
     * @param listener AdListener implementation or null to remove listener
     */
    fun setAdListener(listener: AdListener?) {
        this.listener = listener
    }

    /**
     * Converts density-independent pixels (dp) to actual pixels.
     *
     * @param dp Value in density-independent pixels
     * @return Value in pixels based on device density
     */
    private fun dpToPx(dp: Int): Int {
        return (dp * resources.displayMetrics.density).toInt()
    }

    /**
     * Converts pixels to density-independent pixels (dp).
     *
     * @param px Value in pixels
     * @return Value in density-independent pixels
     */
    private fun pxToDp(px: Int): Int {
        return (px / resources.displayMetrics.density).toInt()
    }

    /**
     * Loads an ad with the specified AdRequest.
     * This is the main entry point for publishers to request and display ads.
     * Note: If an ad is already loading, this call will be ignored.
     *
     * @param adRequest The AdRequest containing ad configuration (test mode, etc.)
     * @throws SecurityException if INTERNET permission is not granted
     */
    @RequiresPermission("android.permission.INTERNET")
    fun loadAd(adRequest: AdRequest) {
        SdkShield.runSafely("BaseAdView.loadAd") {
            if (isLoading) {
                Log.w(TAG, "loadAd ignored - ad is already loading")
                return@runSafely
            }

            if (adUnitId == null || adUnitId.isEmpty()) {
                Log.e(TAG, "Ad unit ID is null or empty")
                listener?.onAdFailedToLoad("Ad unit ID is null or empty")
                return@runSafely
            }

            // Reset destroyed flag to allow reloading
            isDestroyed = false
            isLoading = true

            // Destroy any existing WebView before loading new ad
            if (webView != null) {
                safelyDestroyWebView()
            }

            // Wait a bit before loading new ad to ensure cleanup completes
            mainHandler?.postDelayed({
                if (!isDestroyed) {
                    startAdLoad(adRequest)
                }
            }, 400)
        }
    }

    /**
     * Initiates the actual ad loading process by fetching creative from the server.
     *
     * @param adRequest The AdRequest containing configuration parameters
     */
    private fun startAdLoad(adRequest: AdRequest) {
        try {
            val adgeist = getInstance()
            val fetchCreative: FetchCreative = adgeist.getCreative()

            isTestMode = adRequest.isTestMode

            fetchCreative.fetchCreative(
                adUnitId, "FIXED", isTestMode
            ) { result ->
                mainHandler?.post {
                    if (isDestroyed) return@post
                    isLoading = false

                    if (!result.isSuccess) {
                        Log.e(TAG, "API error: ${result.errorMessage}, statusCode: ${result.statusCode}")
                        listener?.onAdFailedToLoad(result.errorMessage)
                        return@post
                    }

                    try {
                        val campaignDetails = result.data as FixedAdResponse

                        if (campaignDetails.creativesV1.isNullOrEmpty()) {
                            Log.e(TAG, "Empty creative list")
                            listener?.onAdFailedToLoad("Empty creative")
                            return@post
                        }

                        metaData = campaignDetails.metaData
                        val propertiesForAdCard = mutableMapOf<String, Any?>()

                        propertiesForAdCard["adspaceType"] = adType.value
                        propertiesForAdCard["adElementId"] = "adgeist_ads_iframe_$adUnitId"
                        propertiesForAdCard["name"] = campaignDetails.advertiser?.name ?: "-"

                        val options = campaignDetails.displayOptions

                        propertiesForAdCard["isResponsive"] = options?.isResponsive ?: false
                        propertiesForAdCard["responsiveType"] = options?.responsiveType ?: "Square"

                        val creativeDataFromApiResponse = campaignDetails.creativesV1[0]

                        propertiesForAdCard["title"] = creativeDataFromApiResponse.title
                        propertiesForAdCard["description"] = creativeDataFromApiResponse.description
                        propertiesForAdCard["ctaUrl"] = creativeDataFromApiResponse.ctaUrl

                        Log.d(TAG, "measuredWidth: ${pxToDp(measuredWidth)}, measuredHeight: ${pxToDp(measuredHeight)}")
                        if (adIsResponsive) {
                            propertiesForAdCard["width"] = pxToDp(measuredWidth)
                            propertiesForAdCard["height"] = pxToDp(measuredHeight)
                        } else {
                            propertiesForAdCard["width"] = adSize!!.width
                            propertiesForAdCard["height"] = adSize!!.height
                        }

                        // Add primaryCreative
                        val primaryCreative = mutableMapOf<String, String?>()
                        primaryCreative["src"] = creativeDataFromApiResponse.primary?.fileUrl
                        primaryCreative["thumbnailUrl"] = creativeDataFromApiResponse.primary?.thumbnailUrl
                        primaryCreative["type"] = creativeDataFromApiResponse.primary?.type

                        // Add companionCreative
                        val companionCreative = creativeDataFromApiResponse.companions?.map { companion ->
                            mapOf(
                                "src" to companion.fileUrl,
                                "thumbnailUrl" to companion.thumbnailUrl,
                                "type" to companion.type
                            )
                        } ?: emptyList()

                        val mediaList = mutableListOf<Map<String, String?>>()
                        mediaList.add(primaryCreative)
                        mediaList.addAll(companionCreative)
                        propertiesForAdCard["media"] = mediaList

                        val creativeJson = Gson().toJson(propertiesForAdCard)
                        renderAdWithAdCard(creativeJson)
                    } catch (err: Exception) {
                        Log.e(TAG, "Parsing error: ${err.message}", err)
                        listener?.onAdFailedToLoad(err.message ?: "Error")
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "startAdLoad exception: ${e.message}", e)
            listener?.onAdFailedToLoad(e.message ?: "Unknown error")
            mainHandler?.post {
                isLoading = false
            }
        }
    }

    /**
     * Creates and configures a new WebView, sets up JavaScript bridge,
     *
     * @param creativeJsonData JSON string containing creative data for rendering
     */
    private fun renderAdWithAdCard(creativeJsonData: String) {
        if (isDestroyed) return

        removeAllViews()

        webView = WebView(context).apply {
            setBackgroundColor(Color.TRANSPARENT)
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true
            settings.loadWithOverviewMode = true
            settings.useWideViewPort = true
        }

        jsInterface = JsBridge(this, context)
        listener?.onAdOpened()

        // Enable WebView debugging (you can inspect in Chrome DevTools)
        // chrome://inspect/#devices
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(true)
        }

        webView!!.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                return SdkShield.runSafelyWithReturn("BaseAdView.shouldOverrideUrlLoading", false) {
                    openInBrowser(context, url)
                    jsInterface!!.recordClickListener()
                    true
                }
            }

            override fun shouldOverrideUrlLoading(
                view: WebView,
                request: WebResourceRequest
            ): Boolean {
                return SdkShield.runSafelyWithReturn("BaseAdView.shouldOverrideUrlLoading", false) {
                    val url = request.url.toString()
                    openInBrowser(context, url)
                    jsInterface!!.recordClickListener()
                    true
                }
            }

            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                Log.i(TAG, "✅ WebView page finished loading: $url")
            }

            override fun onLoadResource(view: WebView, url: String) {
                super.onLoadResource(view, url)
                Log.d(TAG, "📦 Loading resource: $url")
            }
        }

        // Set WebChromeClient to capture console logs
        webView!!.webChromeClient = object : WebChromeClient() {
            override fun onConsoleMessage(consoleMessage: ConsoleMessage): Boolean {
                val logLevel = consoleMessage.messageLevel().name
                val message = consoleMessage.message()
                val source = consoleMessage.sourceId()
                val line = consoleMessage.lineNumber()

                val fullLog = String.format("[%s] %s (%s:%d)", logLevel, message, source, line)
                when (consoleMessage.messageLevel()) {
                    MessageLevel.ERROR -> Log.e(TAG, "JS Error: $fullLog")
                    MessageLevel.WARNING -> Log.w(TAG, "JS Warning: $fullLog")
                    else -> Log.d(TAG,"🔵 JS Log: $fullLog")
                }
                return true
            }
        }

        // JavaScript bridge interface accessible as 'Android' from WebView HTML
        webView!!.addJavascriptInterface(jsInterface!!, "Android")

        val htmlContent = buildAdCardHtml(creativeJsonData)
        webView!!.loadDataWithBaseURL(
            "https://adgeist.ai",
            htmlContent,
            "text/html",
            "UTF-8",
            null
        )

        // Add WebView to container with full dimensions
        addView(
            webView, LayoutParams(
                LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT
            )
        )

        // Hide companion ads initially until overflow check completes
        if (adType == AdType.COMPANION) {
            webView!!.visibility = View.INVISIBLE
        }
    }

    /**
     * Builds the HTML content for ad rendering in WebView.
     * Loads the main ad view file from assets and injects creative data.
     * This file loads the AdCard library from S3 and renders the ad.
     *
     * @param creativeJsonData JSON string with creative data
     * @return Complete HTML string ready to be loaded in WebView
     */
    private fun buildAdCardHtml(creativeJsonData: String): String {
        val escapedJson = creativeJsonData
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
            .replace("`", "\\`")

        return try {
            val template = context.assets.open("ad_view.html").bufferedReader().use { it.readText() }
            template.replace("{{CREATIVE_DATA}}", escapedJson)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load ad view from assets", e)
            return ""
        }
    }

    /**
     * Positions the child view (WebView) within the ad container.
     * Centers the ad content both horizontally and vertically.
     *
     * @param changed True if this is a different layout than the previous one
     * @param left Left position relative to parent
     * @param top Top position relative to parent
     * @param right Right position relative to parent
     * @param bottom Bottom position relative to parent
     */
    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        val child = getChildAt(0)
        if (child != null && child.visibility != GONE) {
            val width = child.measuredWidth
            val height = child.measuredHeight

            val horizontalSpacing = (right - left - width) / 2
            val verticalSpacing = (bottom - top - height) / 2

            child.layout(
                horizontalSpacing,
                verticalSpacing,
                horizontalSpacing + width,
                verticalSpacing + height
            )
        }
    }

    /**
     * Measures the dimensions of the ad view based on ad size or child view.
     * Handles responsive sizing and fixed ad sizes.
     *
     * @param widthMeasureSpec Width measurement specification from parent
     * @param heightMeasureSpec Height measurement specification from parent
     */
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val child = getChildAt(0)
        var width: Int
        var height: Int

        if (child != null && child.visibility != GONE) {
            measureChild(child, widthMeasureSpec, heightMeasureSpec)
            width = child.measuredWidth
            height = child.measuredHeight
        } else {
            if (adIsResponsive) {
                Log.d(TAG, "Ad is responsive - using available space for measurement")
                width = android.view.View.MeasureSpec.getSize(widthMeasureSpec)
                height = android.view.View.MeasureSpec.getSize(heightMeasureSpec)
            } else if (adSize != null) {
                width = adSize!!.getWidthInPixels(context)
                height = adSize!!.getHeightInPixels(context)
            } else {
                width = 0
                height = 0
            }
        }

        width = max(width.toDouble(), suggestedMinimumWidth.toDouble()).toInt()
        height = max(height.toDouble(), suggestedMinimumHeight.toDouble()).toInt()

        Log.d(TAG, "onMeasure - width: ${resolveSize(width, widthMeasureSpec)}, height: ${resolveSize(height, heightMeasureSpec)}")
        setMeasuredDimension(
            resolveSize(width, widthMeasureSpec),
            resolveSize(height, heightMeasureSpec)
        )
    }

    /**
     * Opens a URL in the device's default browser.
     *
     * @param context Android context for launching intent
     * @param url URL to open in browser
     */
    private fun openInBrowser(context: Context, url: String) {
        try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        } catch (e: Exception) {
            Log.e(
                TAG,
                "Failed to open external URL: $url", e
            )
        }
    }

    /**
     * Called when the view is attached to a window.
     * Resumes WebView rendering if available.
     */
    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        SdkShield.runSafely("BaseAdView.onAttachedToWindow") {
            if (webView != null && !isDestroyed) {
                webView!!.onResume()
            }
        }
    }

    /**
     * Called when window visibility changes.
     * Pauses/resumes WebView rendering to conserve resources.
     *
     * @param visibility New visibility state (VISIBLE, INVISIBLE, or GONE)
     */
    override fun onWindowVisibilityChanged(visibility: Int) {
        super.onWindowVisibilityChanged(visibility)
        SdkShield.runSafely("BaseAdView.onWindowVisibilityChanged") {
            if (webView == null || isDestroyed) return@runSafely

            if (visibility == VISIBLE) {
                webView!!.onResume()
            } else {
                webView!!.onPause()
            }
        }
    }

    /**
     * Called when the view is detached from a window.
     * Triggers WebView cleanup and notifies listener.
     */
    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        SdkShield.runSafely("BaseAdView.onDetachedFromWindow") {
            onDestroyWebView()
        }
    }

    /**
     * Sets the ad dimensions for fixed-size ads.
     *
     * @param adSize The desired ad size
     * @throws IllegalArgumentException if adSize is null
     */
    fun setAdDimension(adSize: AdSize) {
        requireNotNull(adSize) { "AdSize cannot be null" }
        this.adSize = adSize
        requestLayout()
    }

    val isCollapsible: Boolean
        get() = false

    /**
     * Destroys the ad view and releases all resources.
     */
    fun destroy() {
        isLoading = false
        safelyDestroyWebView()
    }

    /**
     * Removes this view from its parent ViewGroup.
     */
    fun removeFromParent() {
        try {
            (parent as? ViewGroup)?.removeView(this)
        } catch (e: Exception) {
            Log.e(TAG, "Error removing from parent: ${e.message}", e)
        }
    }

    /**
     * Safely destroys the WebView to prevent memory leaks.
     * Performs cleanup in the correct order:
     * 1. Removes JavaScript interface
     * 2. Stops loading and pauses rendering
     * 3. Clears history and cache
     * 4. Removes from view hierarchy
     * 5. Loads blank page
     * 6. Destroys WebView instance
     */
    private fun safelyDestroyWebView() {
        if (isDestroyed) return
        isDestroyed = true

        val webViewToDestroy = webView
        webView = null
        jsInterface?.destroyListeners()
        jsInterface = null

        if (webViewToDestroy == null) return

        mainHandler?.post {
            try {
                try {
                    webViewToDestroy.removeJavascriptInterface("Android")
                } catch (e: Exception) { /* ignore */
                }

                webViewToDestroy.stopLoading()
                webViewToDestroy.onPause()
                webViewToDestroy.clearHistory()
                webViewToDestroy.clearCache(true)
                (webViewToDestroy.parent as? ViewGroup)?.removeView(webViewToDestroy)
                removeAllViews()

                try {
                    webViewToDestroy.loadUrl("about:blank")
                } catch (e: Exception) { /* ignore */
                }

                mainHandler?.postDelayed({
                    try {
                        webViewToDestroy.destroy()
                    } catch (e: Exception) {
                        Log.e(TAG, "WebView final destroy failed", e)
                    }
                }, 600)
            } catch (e: Exception) {
                Log.e(TAG, "WebView cleanup error", e)
            }
        }
    }

    /**
     * Internal method called when WebView is being destroyed.
     */
    private fun onDestroyWebView() {
        listener?.onAdClosed()
        safelyDestroyWebView()
    }
}
